feat(previews): allow ffmpeg to connect direct for AWS S3 buckets

invario opened the pull request:
* Resolves: #52079 

## Summary
Allow preview generation on AWS S3 buckets by creating a presigned URL for the file and then passing it directly to ffmpeg which is able to handle http/https already. It can locate the moov atom even if it's at the end of the video file and it does not need to retrieve the whole file to do it.

The presigned URL expires after 10 minute which should be more than enough for this purpose.

## TODO

- [ ] Need more testing.
- [ ] Open to suggestions, critiques, criticisms.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

invario commented:
Updated with a test to see if encryption is enabled on the file, and if so, do not attempt to connect direct.

come-nc reviewed (commented):
Looks good codewise, but needs tests.
Not sure why there is the str_starts_with test for http?

invario commented:
> Looks good codewise, but needs tests. Not sure why there is the str_starts_with test for http?

The test is needed because `getDirectDownload` varies by the external mount provider. SMB, for example, doesn't currently return anything but in the future it might return something like `smb://user@example.domain/foo.bar` Same thing for ftp and potentially others.

invario commented:
Squatted commits and cs fixed.

invario commented:
Pushed again, cs should be good now.

invario committed:
feat(previews): allow ffmpeg to connect direct for AWS S3 buckets

Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: invario <67800603+invario@users.noreply.github.com>

invario commented:
Rebased to 33 and reworked the code into the new `Movie.php`

invario commented on apps/files_external/lib/Lib/Storage/AmazonS3.php:
```diff
@@ -756,4 +757,30 @@ public function writeStream(string $path, $stream, ?int $size = null): int {
+			'Key' => $path,
+		]);
+		// generate a presigned URL that expires after 1 minute
+		$request = $this->getConnection()->createPresignedRequest($command, '+1 minute', []);
```
Increased to 10 minutes for now.

invario commented on apps/files_external/lib/Lib/Storage/AmazonS3.php:
```diff
@@ -756,4 +757,30 @@ public function writeStream(string $path, $stream, ?int $size = null): int {
+		}
+		$result = [
+			'url' => $presignedUrl,
+			'presigned' => true,
```
Timestamp added.  It now returns an additional key named `expiration` that is a `DateTime` object.

CarlSchwan committed:
feat(movie-preview): Use getDirectDownloadById for generating preview

Allow to speed-up considerably the creation of previews for movies
stored on S3.

Signed-off-by: Carl Schwan <carlschwan@kde.org>

CarlSchwan committed:
feat(external-s3): Expose config option to enable direct s3 preview

Signed-off-by: Carl Schwan <carlschwan@kde.org>

CarlSchwan commented:
Rebased and updated to the new interface :)

Tested on both an s3 server mounted as external storage and an s3 server mounted as main storage

icewind1991 commented on lib/private/Preview/Movie.php:
```diff
@@ -42,6 +42,23 @@ public function isAvailable(FileInfo $file): bool {
 	}
 
+	private function connectDirect(File $file): string|false {
+		if ($file->isEncrypted() || stream_get_meta_data($file->fopen('r'))['seekable'] !== true) {
```
What is the reasoning for the `seekable` check?

This is a not insignificant overhead as it does an additional http request and then doesn't use the result

invario replied:
My understanding was that if a stream is not `seekable`, then `fseek` will not work, and from Googling, not all S3 is necessarily seekable.

printminion-co commented:
/backport to stable32
